src/Servers/Kestrel/Core/src/Middleware/LoggingMultiplexedConnectionMiddleware.cs

```diff
@@ -0,0 +1,138 @@
+                }
+                return context;
+            }
+
+            public override async ValueTask<ConnectionContext> ConnectAsync(IFeatureCollection? features = null, CancellationToken cancellationToken = default)
+            {
+                var context = await _inner.ConnectAsync(features, cancellationToken);
+                context = new LoggingConnectionContext(context, _logger);
+                return context;
+            }
+
+            public override CancellationToken ConnectionClosed { get => _inner.ConnectionClosed; set => _inner.ConnectionClosed = value; }
+            public override ValueTask DisposeAsync()
+            {
+                return _inner.DisposeAsync();
+            }
+            public override EndPoint? LocalEndPoint { get => _inner.LocalEndPoint; set => _inner.LocalEndPoint = value; }
+            public override EndPoint? RemoteEndPoint { get => _inner.RemoteEndPoint; set => _inner.RemoteEndPoint = value; }
+        }
+
+        /// <summary>
+        /// Wraps transport with <see cref="LoggingDuplexPipe"/>.
+        /// </summary>
+        private class LoggingConnectionContext : ConnectionContext
+        {
+            private readonly ConnectionContext _inner;
+            private readonly ILogger _logger;
+            private readonly LoggingDuplexPipe _loggingPipe;
+
+            public LoggingConnectionContext(ConnectionContext inner, ILogger logger)
+            {
+                _inner = inner;
+                _logger = logger;
+
+                _loggingPipe = new LoggingDuplexPipe(_inner.Transport, _logger);
+
+                Transport = _loggingPipe;
+            }
+
+            public override void Abort()
+            {
+                _inner.Abort();
+            }
+            public override void Abort(ConnectionAbortedException abortReason)
+            {
+                _inner.Abort(abortReason);
+            }
+
+            public override CancellationToken ConnectionClosed { get => _inner.ConnectionClosed; set => _inner.ConnectionClosed = value; }
+
+            public override string ConnectionId { get => _inner.ConnectionId; set => _inner.ConnectionId = value; }
+
+            public override async ValueTask DisposeAsync()
+            {
+                await _loggingPipe.DisposeAsync();
+                await _inner.DisposeAsync();
+            }
+
+            public override IFeatureCollection Features => _inner.Features;
+
+            public override IDictionary<object, object?> Items { get => _inner.Items; set => _inner.Items = value; }
+
+            public override EndPoint? LocalEndPoint { get => _inner.LocalEndPoint; set => _inner.LocalEndPoint = value; }
+            public override EndPoint? RemoteEndPoint { get => _inner.RemoteEndPoint; set => _inner.RemoteEndPoint = value; }
+
+            // Note that the inner transport isn't replaced so it is still completed on dispose
```

Comment:
This comment should no longer be necessary now that https://github.com/dotnet/aspnetcore/pull/30416 is merged.